src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/BitVector.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+
+namespace System.Text.RegularExpressions.Symbolic
+{
+    /// <summary>Represents an immutable bit vector of an arbitrary number of bits.</summary>
+    /// <remarks>
+    /// This differs from <see cref="BitArray"/> in that it's a value type rather than a reference type
+    /// and it exposes the ability to quickly compute equality and a comparison.
+    /// </remarks>
+    internal struct BitVector : IComparable<BitVector>, IEquatable<BitVector>
+    {
+        /// <summary>Stores the bits in an array of 64-bit integers.</summary>
+        /// <remarks>
+        /// If Length is not evenly divisible by 64 then the remainingbits are
+        /// in the least significant bits of the last element.
+        /// </remarks>
+        private readonly ulong[] _blocks;
+        /// <summary>The number of bits represented by the BitVector.</summary>
+        public readonly int Length;
+        /// <summary>Lazily-computed hash code value.</summary>
+        private int? _hashcode;
+
+        /// <summary>Initializes the <see cref="BitVector"/> to be of the specified length and contain all false bits.</summary>
+        private BitVector(int length)
```

Comment:
NIT: should we have some code ensuring length is > than 0?

Reply:
Sure, I've added an assert.